blessed/sequences.py

```diff
@@ -315,6 +315,28 @@ def center(self, width, fillchar=u' '):
             (max(0.0, math.ceil(split))) / float(len(fillchar)))
         return u''.join((leftside, self, rightside))
 
+    def truncate(self, width):
+        """
+        Truncate a string in a sequence-aware manner.
+
+        Any printable characters beyond ``width`` are removed, while all
+        sequences remain in place.
+
+        :arg int width: The printable width to truncate the string to.
+        :rtype: str
+        :returns: String truncated to at most ``width`` printable characters.
+        """
+        output = ""
+        current_width = 0
+        for text, cap in iter_parse(self._term, self):
+            if not cap:
```

Comment:
Should this be `if not cap and current_width < width:`? This would save calling `wcwidth()` unnecessarily.

Reply:
No, otherwise you end up in a situation where a wide character may not get truncated when it should. Doesn't work with wide printing characters.

Maybe we could do something with the generator, though. Something like:
```python
        output = ""
        current_width = 0
        iter = iter_parse(self._tern, self)
        try:
            while current_width < width:
                text, cap = next(iter)
                if not cap:
                    current_width += wcwidth(text)
                    if current_width <= width:
                        output += text
                    continue
                output += text
            for text, cap in iter:
                if cap:
                    output += text
        except StopIteration:
            pass
        return output
```
To at least not have to call `wcwidth` when we know we won't need it any more, although the way I've written that there, you now have to check twice. Conditionals are mostly free, but it's something. It's also a lot uglier and a little harder to understand. Not sure about Python2 compatibility either, I know there was something about the `next` in iterations changing or something.

All in all, I think it's okay the way it is, but I can change it if you like.

Reply:
I don't exactly follow your logic, but I'm guessing you're concerned there will be zero width characters at the truncation boundary? I don't really know what effect that has on Unicode. I also don't know what affect we have as it's written now including all the zero width characters while removing the characters between them. @jquast probably has a better idea how that behaves.

Reply:
Thought about it a little more, maybe something like this would be better
```python
output = ""
current_width = 0
it = iter_parse(self._term, self)
for text, cap in it:
    if not cap:
        current_width += wcwidth(text)
        if current_width <= width:
            output += text
        else:
            break
        continue
    output += text
for text, cap in it:
    if cap:
        output += text
return output
```

Assuming the generator works the way I expect it to.

Reply:
I don't think that would work. Not sure why'd want to iterate twice and on the second iteration would be empty already. This also wouldn't address running `wcwidth()` more than necessary.
Now that I look at it, you could simplify your original logic a little though. Still doesn't address the `wcwidth()` thing, but a little better.

``` python
        for text, cap in iter_parse(self._term, self):
            if not cap:
                current_width += wcwidth(text)
                if current_width > width:
                    continue
            output += text
        return output

Reply:
For the logic, consider the following:

Say you have a string, `"ABＣ"` you want to truncate to 3 characters. 

The Ｃ is a full width character, and thus has a width of 2, so it should be truncated if you were to call `Terminal.truncate("ABＣ", 3)`. 

If you switch the conditional to `if not cap and current_width < width:`, it would not be truncated, but rather included in the string. This matters for things like if you want to truncate to say, the width of the terminal before printing, so you don't end up with the terminal wrapping it for you and breaking your lines. It's a corner case since it's only a problem if you have a wide character at the very end of something you want to truncate, but the correct behaviour is to truncate.

But I think my second iteration of it solves both issues in a decent way.

I did, however, run into an issue with Python 2 when trying to test it.

```python
def test_truncate_wide_end(all_terms):
    """Ensure that terminal.truncate has the correct behaviour for wide characters."""
    @as_subprocess
    def child(kind):
        from blessed import Terminal
        term = Terminal(kind)
        test_string = "ABＣ"
        assert term.truncate(test_string, 3) == "AB"

    child(all_terms)
```

This is the test I made. Works fine on Python3, but on Python2, I get the output

```console
E       AssertionError: Output in child process:
E       ========================================
E         File "/home/alligator/prog/sources/blessed/tests/accessories.py", line 81, in __call__
E           self.func(*args, **kwargs)
E         File "/home/alligator/prog/sources/blessed/tests/test_sequences.py", line 684, in child
E           assert term.truncate(test_string, 3) == "AB"
E         File "/home/alligator/prog/sources/blessed/blessed/terminal.py", line 1048, in truncate
E           return Sequence(text, self).truncate(width)
E         File "/home/alligator/prog/sources/blessed/blessed/sequences.py", line 269, in __new__
E           new = six.text_type.__new__(cls, sequence_text)
E       -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 2: ordinal not in range(128)
E       ========================================

tests/accessories.py:133: AssertionError
```

Which seems to be an issue with the `Sequence` constructor not being happy with wide characters. Any ideas on how I should work around that?

Reply:
`test_string = u"ABＣ"` instead of `test_string = "ABＣ"`

Reply:
Ah, of course. Thank you, I've never really done any Python2 stuff.

Anyway, I updated the implementation in a974f05

```python
output = ""
current_width = 0
iterator = iter_parse(self._term, self)
for text, cap in iterator:
    if not cap:
        current_width += wcwidth(text)
        if current_width > width:
            break
    output += text
for text, cap in iterator:
    if cap:
        output += text
return output
```

The reason this works is because we save the generator (and its state) so we can iterate over it with a different algorithm depending. That's what the `break` statement there does.

First we simply go through each element in the text, if it's printable we check the width. If the width is okay, we add it.

If the width is not okay, we break out of the first iteration loop and proceed to the second. In the second we simply check if it's a control sequence, since we know for sure we never want any more text after that, and we can skip the `wcwidth` call.

In the worst case, where the entire text fits into the width after all, the generator is already fully exhausted, so tthe second loop is skipped naturally.

Reply:
Ok, it was clearer in your last commit. I modified it for readability.

@jquast, I'm happy with this once we rebase for testing, but this is different from the original and drops 0-length unicode from the rest of the string once we reach the desired width. I don't know if that's what we want or not. What do you think?

Reply:
wcwidth is difficult to use correctly, I maintain that library too, I will commit a change and test, sigh!

I’d really like to work on wcwidth, to make it efficient C code and also Provide a  new API of a more basic “length” function or something that doesn’t require “check for -1 return value” which is legacy from C Posix API..

Reply:
@jquast I cloned his fork and committed there. I think when you create a pull request it will give maintainers commit access to your source branch by default.

Reply:
thanks, I didn't know that!